### DISCUSSION TITLE
feature: a way to override nvim-cmp to use a fork

### DISCUSSION THREAD

**Discussion by OneOfOne** [LazyVim/LazyVim#4549] [Q&A] [answered]
### Did you check the docs?

- [x] I have read all the LazyVim docs
- [x] This is not a request for a new or existing extra (see above)

### Is your feature request related to a problem? Please describe.

Right now there's no way to override nvim-cmp to use a fork.

### Describe the solution you'd like

The current solution is ugly:

```lua
	{
		'iguanacucumber/magazine.nvim',
		name = 'nvim-cmp',
		dev = true,
		build = 'git clone https://github.com/iguanacucumber/magazine.nvim ~/projects/nvim-cmp/',
..........
```
Which also requires updating ~/projects/nvim-cmp by hand.

### Describe alternatives you've considered

Would have to edit all the extra that uses it to use the new fork instead.

**Comment by dpetka2001** [comment]
Honestly the naming of this plugin is just not good with regard to Neovim distros. He should have forked the repo and make all these additions under a separate branch like the `yioneko` fork. Then users can just do
```lua
{
		'yioneko/nvim-cmp',
		branch = "perf-up",
}
```
That way the `nvim-cmp` plugin spec will remain the same without other changes. The current naming of the plugin is only friendly for custom configurations from scratch in my limited knowledge and as you say you would have to change the naming in other extras as well, because it's simply a different plugin spec. Consider making a feature request at the plugin's repo, so that the author takes into consideration changing the name of this plugin to just his forked `nvim-cmp` repo. Something like `iguanacucumber/nvim-cmp` with the changes under a separate branch.

**Reply by OneOfOne** [reply]
that doesn't work either.

**Reply by OneOfOne** [reply]
name = 'nvim-cmp', fixes the different plugin name issue, however it doesn't fix the core issue that it tries to download the original regardless of what you put without `dev = true`.

**Comment by OneOfOne** [comment]
```lua
require('lazy').setup({
	spec = {
		{
			'iguanacucumber/magazine.nvim',
			name = 'nvim-cmp',
			dev = true,
			build = 'git clone https://github.com/iguanacucumber/magazine.nvim ~/projects/nvim-cmp/',
		},
		{ 'LazyVim/LazyVim', import = 'lazyvim.plugins' },
,,,,,,,,,
})
```
works

**Reply by dpetka2001** [reply]
Thank you for pointing me to the `name` attribute. I remember I had seen it before but did not totally understood what it does. How I installed it without cloning the repo locally.
1. Open Neovim
2. Create a file `~/.config/nvim/lua/plugins/magazine.lua` with contents
```lua
return {
  {
    "iguanacucumber/magazine.nvim",
    name = "nvim-cmp",
  },
}
```
3. Save the file and open Lazy UI with `<leader>l` (lowercase L)
4. Press `C` for checking for updates and then `U` for the updates to take place. It should also update the remote for `nvim-cmp` plugin spec.
5. Restart Neovim

![2024-10-08_10-33](https://github.com/user-attachments/assets/5f953ba7-f1c2-4274-9219-3dbf0df1396e)

This is what I get. I also installed the Tailwind Extra and it doesn't download the old plugin spec with the old remote.

**Reply by OneOfOne** [reply]
that doesn't work for me:

![Screenshot_20241008_123713](https://github.com/user-attachments/assets/ee5eb718-5338-4691-90b9-6404194f2bab)

**Reply by OneOfOne** [reply]
I got it working by adding it as a dependency of another plugin:

```lua
	{
		'nvim-telescope/telescope.nvim',
		dependencies = {
			{
				'iguanacucumber/magazine.nvim',
				name = 'nvim-cmp',
				-- dev = true,
				-- build = 'git clone https://github.com/iguanacucumber/magazine.nvim ~/projects/nvim-cmp/',
			},
       }
```

**Reply by dpetka2001** [reply]
I did it exactly as I mentioned in my previous post and I'm also able to get the updates from today via Lazy.

**Reply by dpetka2001** [reply, marked as answer]
Screencast with the steps from a LazyVim installation from scratch.

[Screencast_2024-10-08-21-20-34.webm](https://github.com/user-attachments/assets/af750888-3e6d-4b1d-ac42-b881181c64b8)

**Reply by OneOfOne** [reply]
I''m not sure why that didn't work at all for me, but I'll mark yours as the answer.

**Reply by dpetka2001** [reply]
I just tried your dotfiles. It's because you have `nvim-cmp` listed as a dependency of `nvim-lspconfig` in file `code.lua`. If you comment out that part (it's not really needed in my opinion), then if you make the change in `cmp.lua` file it will take effect with the steps I mentioned. 

The reason for this is that `lazy.nvim` parses the files alphabetically. And the plugin spec that gets parsed last, is the one that should have the change from `nvim-cmp` to `magazine.nvim`. So, because file `code.lua` will get parsed after `cmp.lua` the dependency that you have there for `hrsh7th/nvim-cmp` is taking over the change that you make in `cmp.lua` file. Try it to see for yourself.

**Reply by OneOfOne** [reply]
That worked, thank you very much!
